driver/interrupts.c

```diff
@@ -78,15 +101,12 @@ wait_queue_head_t *interrupts_wait_queue(struct interrupt_control *control)
 /* Stores user space interrupt events and notifies as appropriate. */
 static void event_interrupt(struct interrupt_control *control, uint32_t events)
 {
-    /* Add the new events into the current event mask. */
-    int old_events;
-    do
-        old_events = atomic_read(&control->events);
-    while (atomic_cmpxchg(
-        &control->events, old_events, old_events | events) != old_events);
-
+    /* Add the new events into the current event masks. */
+    for (int reader=0; reader < N_EVENT_READERS; reader++)
+         atomic_set(&control->events[reader],
```

Comment:
@Araneidae , I think we still need a spinlock to protect the event mask, consider the following scenario: `atomic_read` gets the old mask value, at the same time, some user thread is reading the mask and setting it to 0 (this happens as consequence of an older interrupt), then `atomic_set` writes the new value and the user gets some event/s twice the next time he's awaken.

Reply:
This should be `atomic_or` as far as I can tell.

Otherwise, I don't see anything useful that a spinlock can do for us.

Reply:
Oh, I see.  When I comment on the same bit of code it becomes a reply here.  Anyhow, I think using `atomic_or` should alleviate your concerns.

Reply:
fixed! it builds on my machine but we'll need to check if the target kernel has the `atomic_or` function available